chainerx_cc/chainerx/cuda/cuda_device/std_ops.h

```diff
@@ -0,0 +1,60 @@
+
+#define CHAINERX_CUDA_REGISTER_ELTWISE_DTYPE_UNARY_OP(func, func_def, visit_dtype)      \
+                                                                                        \
+    template <typename T>                                                               \
+    struct func##Impl {                                                                 \
+        using CudaType = cuda_internal::DataType<T>;                                    \
+        __device__ void operator()(int64_t i, CudaType x, CudaType& out) func_def       \
+    };                                                                                  \
+                                                                                        \
+    class Cuda##func##Op : public func##Op {                                            \
+    public:                                                                             \
+        void Call(const Array& x, const Array& out) override {                          \
+            Device& device = x.device();                                                \
+            device.CheckDevicesCompatible(x, out);                                      \
+            CudaSetDeviceScope scope{device.index()};                                   \
+            const Array& x_cast = x.dtype() == out.dtype() ? x : x.AsType(out.dtype()); \
+            visit_dtype(out.dtype(), [&](auto pt) {                                     \
+                using T = typename decltype(pt)::type;                                  \
+                Elementwise<const T, T>(func##Impl<T>{}, x_cast, out);                  \
+            });                                                                         \
+        }                                                                               \
+    };                                                                                  \
+                                                                                        \
+    CHAINERX_REGISTER_OP_CUDA(func##Op, Cuda##func##Op)
+
+#define CHAINERX_CUDA_REGISTER_ELTWISE_FLOAT_UNARY_OP(func, func_def) \
+    CHAINERX_CUDA_REGISTER_ELTWISE_DTYPE_UNARY_OP(func, func_def, VisitFloatingPointDtype)
+
+#define CHAINERX_CUDA_REGISTER_ELTWISE_UNARY_OP(func, func_def) CHAINERX_CUDA_REGISTER_ELTWISE_DTYPE_UNARY_OP(func, func_def, VisitDtype)
+
+#define CHAINERX_CUDA_REGISTER_ELTWISE_DTYPE_BINARY_OP(func, func_def, visit_dtype)                 \
+                                                                                                    \
+    template <typename T>                                                                           \
+    struct func##Impl {                                                                             \
+        using CudaType = cuda_internal::DataType<T>;                                                \
+        __device__ void operator()(int64_t /*i*/, CudaType x1, CudaType x2, CudaType& out) func_def \
```

Comment:
Yeah, I was planning for that actually. Yesterday I was getting `unused-variable` error. Will have a new to have a new macro with index available. Sounds right?

Reply:
I'm wondering if it'd be possible with this macros https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/macro.h#L15-L17 🤔 .

Reply:
I tried `void operator()(int64_t i, T x, T& out) { CHAINERX_ASSERT(func_def); }` . But compiler still complains.

I don't think it will work as I also tried a minimal version [here](https://tio.run/##XU9LS8NAEL7vrxgohF1ogr02aSFI6UmUBPUgsqy7k7qQbMI@ihDy1415HLTOZZjvyciuiy9SjuNGG1kHhZDp1nmLojmSX0w6r2r9cQM1wn/@Ba4ofWuPhGwUVtoglM9Pxaks@UN@XzzSJEkY0GurFaNv79CDRR@sgUrUDlMYKIMoAsr5S87z4lxyzhghsx6CCQ4Vd6Gz6BzVxsPXFqYVQRs86wlMc9vWzwwcYHc3RbOUDITMtkZoQxlZHTMwqdLl@Nex2y7RKSzk9P1@L@fELFstwzh@y6oWFzfGr2htaw9rQtwJKxr0aH8A) which doesn't compile. Let me know if I am doing something wrong.

Reply:
https://stackoverflow.com/a/20077422/5602957

How about the second method using unnamed parameter in the template. I tried that and does work.

Reply:
Ah I meant to pass the index as the argument to `CHAINERX_ASSERT`. Something along the lines of the following inside the macro definition.
```cpp
{
    CHAINERX_ASSERT(i);
    func_def;
}
```

Reply:
As per the stackoverflow answer. 
```
template <class T>
void ignore_unused(T&) {} 

//later...
auto x = /* ... */;
ignore_unused(x);
```
This is better than `(void)i`.
Your thoughts?

I am unable to find the other link which neatly explained why this is better comparted to `void` casting. Will share the link once I find it.

Reply:
I see. The point is that we already have `(void)variable;` occurrences in our source code so it's consistent and fine for this PR. However, if there is a general better approach, we might consider applying that to all (or wherever possible) of the occurrence.

Reply:
Have used `(void)i`.

I was also thinking about allowing the implementer to also be able to cast the based on `input and output` variable.

Maybe pass in a lambda that takes in input and output and return casted version of both?
For example on 
```
[](auto x, auto output) -> std::tuple<Array, Array> {
    // cast and return
};
```

Your thoughts?

Reply:
Do you have concrete cases where this applies? We briefly discussed this offline but thought that it could wait.